Improve nuget.exe restore error message when passing file globs

### NuGet Product Used

NuGet.exe

### Product Version

6.6.1.2

### Worked before?

_No response_

### Impact

It's more difficult to complete my work

### Repro Steps & Context

nuget restore **/*.sln gives error "Illegal characters in path"

here is the list of PATH variable list in command prompt shell


C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\\Extensions\Microsoft\IntelliCode\CLI;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\VC\VCPackages;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\TestWindow;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\bin\Roslyn;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Team Tools\Performance Tools;
C:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTools\vs2019\;
C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools\devinit;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\\MSBuild\Current\Bin;
C:\Windows\Microsoft.NET\Framework\v4.0.30319;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools\;
C:\Program Files\Microsoft\jdk-11.0.16.101-hotspot\bin;
C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;
C:\Windows\system32;
C:\Windows;
C:\Windows\System32\Wbem;
C:\Windows\System32\WindowsPowerShell\v1.0\;
C:\Windows\System32\OpenSSH\;
C:\Program Files (x86)\Citrix\HDX\bin\;
C:\Program Files\Citrix\HDX\bin\;
C:\Program Files\Git\cmd;
C:\Program Files\dotnet\;
C:\ProgramData\chocolatey\bin;
C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;
C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;
C:\Program Files\Microsoft SQL Server\150\Tools\Binn\;
C:\Program Files\SQL Anywhere 17\bin64;
C:\Program Files\SQL Anywhere 17\bin32;
C:\Program Files (x86)\Appeon\Shared\PowerBuilder;
C:\Program Files (x86)\Appeon\Shared\PowerBuilder\x64;
C:\Program Files (x86)\Appeon\PowerBuilder 17.0;
C:\Windows\Microsoft.NET\Framework\v4.0.30319;
C:\Program Files (x86)\Yarn\bin\;
C:\ProgramData\nvm;
C:\Program Files\nodejs;
C:\Program Files\Microsoft VS Code\bin;
C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn\;
C:\Program Files (x86)\nant-0.92\bin;
C:\Program Files\PowerShell\7\;
C:\Program Files\Azure Data Studio\bin;
C:\Program Files\Microsoft SQL Server\150\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\150\Tools\Binn\;
C:\Users\David.Wei\AppData\Local\Microsoft\WindowsApps;
C:\Users\David.Wei\.dotnet\tools;
C:\Users\David.Wei\AppData\Roaming\npm;
C:\Users\David.Wei\AppData\Local\Yarn\bin;
C:\ProgramData\nvm;
C:\Program Files\nodejs;
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\VC\Linux\bin\ConnectionManagerExe

### Verbose Logs

```shell
nuget restore **/*.sln -verbosity detail gives following details information


NuGet Version: 6.6.1.2
Illegal characters in path.
System.ArgumentException: Illegal characters in path.
   at System.Security.Permissions.FileIOPermission.EmulateFileIOPermissionChecks(String fullPath)
   at System.Security.Permissions.FileIOPermission.QuickDemand(FileIOPermissionAccess access, String fullPath, Boolean checkForDuplicates, Boolean needFullPath)
   at NuGet.CommandLine.RestoreCommand.<DetermineRestoreInputsAsync>d__62.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<ExecuteCommandAsync>d__52.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
```

@vitesse the error message needs improvement, but NuGet.exe doesn't support file globbing, or multiple inputs to a restore. If you have multiple solution files, you need to either script it yourself, or if the number and names of solutions don't change very often, you could hardcode the multiple invocations in your build scripts.

Also, I highly recommend not using nuget.exe for restore. It has some problems that don't exist if you restore via MSBuild instead.  If all the projects in your solution are "SDK style", then you can use `dotnet restore`. Otherwise, use `msbuild -t:restore`. If your solution has at least one project using `packages.config`, then use `msbuild -t:restore -p:RestorePackagesConfig=true`.
For reference, this the output for `dotnet restore **\*.sln`:

For reference, here's MSBuild's output trying to pass in a glob. We can use it as inspiration for a better error message for nuget.exe

```console
> dotnet restore **/*.sln
MSBUILD : error MSB1009: Project file does not exist.
Switch: **/*.sln
```

--- a/src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs
+++ b/src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs
@@ -498,7 +498,22 @@ namespace NuGet.CommandLine
             {
                 // Restore takes multiple arguments, each could be a file or directory
                 var argument = Arguments.Single();
-                var fullPath = Path.GetFullPath(argument);
+                string fullPath;
+                try
+                {
+                    fullPath = Path.GetFullPath(argument);
+                }
+                catch (ArgumentException)
+                {
+                    // Treat "invalid characters in path" like a file not found.
+                    // Afterall, a filename with invalid characters can't exist.
+                    var message = string.Format(
+                        CultureInfo.CurrentCulture,
+                        LocalizedResourceManager.GetString("RestoreCommandFileNotFound"),
+                        argument);
+
+                    throw new InvalidOperationException(message);
+                }
 
                 if (Directory.Exists(fullPath))
                 {

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs
@@ -32,14 +32,15 @@ namespace NuGet.CommandLine.Test
         private const int _failureCode = 1;
         private const int _successCode = 0;
 
-        [Fact]
-        public void RestoreCommand_BadInputPath()
+        [Theory]
+        [InlineData("bad/pat.h/myfile.blah")]
+        [InlineData("**/*.sln")]
+        public void RestoreCommand_BadInputPath(string solutionPath)
         {
             using (var randomTestFolder = TestDirectory.Create())
             {
                 // Arrange
                 var nugetexe = Util.GetNuGetExePath();
-                var solutionPath = "bad/pat.h/myfile.blah";
 
                 var args = new string[]
                 {
@@ -59,7 +60,7 @@ namespace NuGet.CommandLine.Test
                 // Assert
                 Assert.NotEqual(_successCode, r.ExitCode);
                 var error = r.Errors;
-                Assert.Contains("Input file does not exist: bad/pat.h/myfile.blah", r.Errors, StringComparison.OrdinalIgnoreCase);
+                Assert.Contains("Input file does not exist: " + solutionPath, r.Errors, StringComparison.OrdinalIgnoreCase);
             }
         }
 

